downstairs/src/region.rs

```diff
@@ -965,6 +1106,162 @@ impl Region {
+     * 12. Delete completed dir.
+     *
+     *  This also requires the following behavior on every extent open:
+     *   A. If xxx.Copy directory found, delete it.
+     *   B. If xxx.Completed directory found, delete it.
+     *   C. If xxx.Replace dir found start at step 4 above and continue
+     *      on through 6.
+     *   D. Only then, open extent.
+     */
+    pub async fn repair_extent(
+        &mut self,
+        eid: usize,
+        repair_addr: SocketAddr,
+    ) -> Result<(), CrucibleError> {
+        // Make sure the extent:
+        // is currently closed, matches our eid, is not read-only
+        assert!(self.extents[eid].inner.is_none());
+        assert_eq!(self.extents[eid].number, eid as u32);
+        assert!(!self.read_only);
+
+        self.get_extent_copy(eid, repair_addr).await?;
+
+        // Returning from get_extent_copy means we have copied all our
+        // files and moved the copy directory to replace directory.
+        // Now, replace the current extent files with the replacement ones.
+        move_replacement_extent(&self.dir, eid)?;
+
+        Ok(())
+    }
+
+    /**
+     * Connect to the source and pull over all the extent files for the
+     * given extent ID.
+     * The files are loaded into the copy_dir for the given extent.
+     * After all the files have been copied locally, we rename the
+     * copy_dir to replace_dir.
+     */
+    pub async fn get_extent_copy(
+        &mut self,
+        eid: usize,
+        repair_addr: SocketAddr,
+    ) -> Result<(), CrucibleError> {
+        assert!(self.extents[eid].inner.is_none());
+        // Make sure copy, replace, cleanup dirs don't exist yet.
+        // We don't need them yet, but if they exist, then something
+        // is wrong.
+        let rd = replace_dir(&self.dir, eid as u32);
+        if rd.exists() {
+            crucible_bail!(
+                IoError,
+                "Replace directory: {:?} already exists",
+                rd,
+            );
+        }
+
+        let extent = &self.extents[eid];
+        let copy_dir = extent.create_copy_dir(&self.dir)?;
+
+        // XXX TLS someday?  Authentication?
+        let url = format!("http://{:?}", repair_addr);
+        let repair_server = Client::new(&url);
+
+        let repair_files = repair_server
+            .get_files_for_extent(eid as u32)
+            .await
+            .unwrap()
+            .into_inner()
+            .files;
+
+        println!("Found repair files: {:?}", repair_files);
+
+        // The repair file list should always contain the extent data
+        // file itself, and the .db file (metadata) for that extent.
+        // Missing these means the repair will not succeed.
+        let filename = extent_file_name(eid as u32, None);
+        if !repair_files.contains(&filename) {
+            // XXX Panic now, but this will eventually bail and abort
+            // the repair and let the upper layers handle it.
```

Comment:
Yes, sort of.  For some reason the remote downstairs returned a list of files and the generic extent default filename was not one of them.  It could be because the repair source was bogus, but could also be some other failure.  Basically, this repair cannot continue and we are `\o/` throwing up our hands and giving up.  I don't expect to see it during normal operations, and it should not panic, but until I know things are all working, I want to panic as soon as we encounter unexpected behavior.

Reply:
I re-did this with a new function that validates the file list.